### ISSUE TITLE
Phase 41.1 — AdversarialAttackSimulator: FGSM, PGD, C&W, AutoAttack Attack Suite

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#825]
# Phase 41.1 — AdversarialAttackSimulator

## Overview

The **AdversarialAttackSimulator** is the red-team engine of ASI-Build's adversarial robustness pipeline. It implements a comprehensive suite of adversarial attack methods — from fast gradient-based single-step attacks (FGSM) to sophisticated optimization-based attacks (C&W) and ensemble evaluation (AutoAttack) — enabling systematic vulnerability assessment of any model in the ASI-Build ecosystem.

## Architecture

```
┌─────────────────────────────────────────────────────────────┐
│                 AdversarialAttackSimulator                    │
│                                                              │
│  ┌──────────────┐  ┌──────────────┐  ┌──────────────────┐  │
│  │  AttackRegistry│  │ PerturbationBudget│ │ AttackScheduler │  │
│  │              │  │  Manager     │  │                  │  │
│  │  FGSM        │  │  L∞ / L2 /  │  │  Sequential /    │  │
│  │  PGD         │  │  L0 / L1    │  │  Parallel /      │  │
│  │  C&W         │  │  constraints │  │  Adaptive        │  │
│  │  AutoAttack  │  │              │  │                  │  │
│  │  DeepFool    │  │  Per-sample  │  │  Early stopping  │  │
│  │  JSMA        │  │  budgets     │  │  Success tracking│  │
│  └──────┬───────┘  └──────┬───────┘  └────────┬─────────┘  │
│         │                 │                    │             │
│  ┌──────▼─────────────────▼────────────────────▼─────────┐  │
│  │              Attack Execution Engine                    │  │
│  │                                                        │  │
│  │  ┌─────────────┐ ┌──────────────┐ ┌────────────────┐  │  │
│  │  │ Gradient    │ │ Optimization │ │ Score-based    │  │  │
│  │  │ Attacks     │ │ Attacks      │ │ Attacks        │  │  │
│  │  │ (white-box) │ │ (white-box)  │ │ (black-box)    │  │  │
│  │  └─────────────┘ └──────────────┘ └────────────────┘  │  │
│  └────────────────────────┬───────────────────────────────┘  │
│                           │                                  │
│  ┌────────────────────────▼───────────────────────────────┐  │
│  │              Adversarial Example Store                   │  │
│  │   Perturbations • Success rates • Transferability       │  │
│  └─────────────────────────────────────────────────────────┘  │
└─────────────────────────────────────────────────────────────┘
```

## TypeScript Interface

```typescript
import { Tensor, Model } from '@asi-build/core';

/** Norm type for perturbation budget */
type NormType = 'Linf' | 'L2' | 'L1' | 'L0';

/** Attack method identifier */
type AttackMethod = 'fgsm' | 'pgd' | 'cw' | 'deepfool' | 'jsma' | 'autoattack' | 'square' | 'fab';

interface AttackConfig {
  method: AttackMethod;
  norm: NormType;
  epsilon: number;              // perturbation budget
  stepSize?: number;            // step size for iterative attacks
  numSteps?: number;            // iterations for PGD/C&W
  confidence?: number;          // C&W confidence parameter κ
  targeted?: boolean;           // targeted vs untargeted
  targetClass?: number;         // target class for targeted attacks
  randomStart?: boolean;        // random initialization (PGD)
  numRestarts?: number;         // number of random restarts
  batchSize?: number;           // batch processing size
  earlyStop?: boolean;          // stop on first success
}

interface AttackResult {
  original: Tensor;
  adversarial: Tensor;
  perturbation: Tensor;
  originalPrediction: number;
  adversarialPrediction: number;
  success: boolean;
  perturbationNorm: number;
  confidence: number;
  numIterations: number;
  timeMs: number;
}

interface RobustnessReport {
  modelId: string;
  attackMethod: AttackMethod;
  norm: NormType;
  epsilon: number;
  totalSamples: number;
  successfulAttacks: number;
  robustAccuracy: number;       // 1 - attack success rate
  avgPerturbationNorm: number;
  medianPerturbationNorm: number;
  avgIterations: number;
  avgTimeMs: number;
  perClassAccuracy: Map<number, number>;
}

interface AdversarialAttackSimulator {
  // Core attack methods
  fgsm(model: Model, input: Tensor, label: number, epsilon: number): Promise<AttackResult>;
  pgd(model: Model, input: Tensor, label: number, config: AttackConfig): Promise<AttackResult>;
  carliniWagner(model: Model, input: Tensor, label: number, config: AttackConfig): Promise<AttackResult>;
  autoAttack(model: Model, input: Tensor, label: number, epsilon: number): Promise<AttackResult>;
  deepFool(model: Model, input: Tensor, label: number): Promise<AttackResult>;

  // Batch evaluation
  evaluateRobustness(model: Model, dataset: Tensor[], labels: number[], config: AttackConfig): Promise<RobustnessReport>;
  
  // Transferability analysis
  transferAttack(sourceModel: Model, targetModel: Model, input: Tensor, config: AttackConfig): Promise<AttackResult>;
  
  // Attack scheduling
  adaptiveAttack(model: Model, input: Tensor, label: number, methods?: AttackMethod[]): Promise<AttackResult>;
  
  // Storage
  saveAdversarialExamples(results: AttackResult[], path: string): Promise<void>;
  loadAdversarialExamples(path: string): Promise<AttackResult[]>;
}
```

## Configuration

```yaml
adversarial_attack_simulator:
  default_norm: "Linf"
  default_epsilon: 0.031373     # 8/255 for image models
  
  fgsm:
    clip_min: 0.0
    clip_max: 1.0
  
  pgd:
    num_steps: 40
    step_size: 0.007843          # 2/255
    random_start: true
    num_restarts: 10
  
  carlini_wagner:
    confidence: 0.0
    learning_rate: 0.01
    binary_search_steps: 9
    max_iterations: 10000
    initial_const: 0.001
    abort_early: true
  
  autoattack:
    attacks: ["apgd-ce", "apgd-dlr", "fab", "square"]
    version: "standard"         # standard | plus | rand
    
  evaluation:
    batch_size: 128
    num_workers: 4
    save_adversarial: true
    log_interval: 100
  
  transferability:
    ensemble_size: 5
    ensemble_weights: "uniform"
```

### FGSM (Goodfellow et al., 2015)
```python
def fgsm_attack(model, x, y, epsilon):
    """Fast Gradient Sign Method - single-step attack."""
    x_adv = x.clone().requires_grad_(True)
    loss = F.cross_entropy(model(x_adv), y)
    loss.backward()
    perturbation = epsilon * x_adv.grad.sign()
    x_adv = torch.clamp(x + perturbation, 0, 1)
    return x_adv
```

### PGD (Madry et al., 2018)
```python
def pgd_attack(model, x, y, epsilon, step_size, num_steps, random_start=True):
    """Projected Gradient Descent - iterative multi-step attack."""
    x_adv = x.clone()
    if random_start:
        x_adv = x_adv + torch.empty_like(x_adv).uniform_(-epsilon, epsilon)
        x_adv = torch.clamp(x_adv, 0, 1)
    
    for _ in range(num_steps):
        x_adv.requires_grad_(True)
        loss = F.cross_entropy(model(x_adv), y)
        loss.backward()
        with torch.no_grad():
            x_adv = x_adv + step_size * x_adv.grad.sign()
            # Project back to epsilon-ball
            perturbation = torch.clamp(x_adv - x, -epsilon, epsilon)
            x_adv = torch.clamp(x + perturbation, 0, 1)
    return x_adv
```

### C&W (Carlini & Wagner, 2017)
```python
def cw_attack(model, x, y, confidence=0, lr=0.01, max_iter=10000):
    """Carlini & Wagner L2 attack via optimization."""
    w = torch.zeros_like(x, requires_grad=True)
    optimizer = torch.optim.Adam([w], lr=lr)
    best_adv = x.clone()
    best_dist = float('inf')
    
    for _ in range(max_iter):
        x_adv = 0.5 * (torch.tanh(w) + 1)  # box constraint via tanh
        logits = model(x_adv)
        
        # f(x') = max(Z(x')_y - max_{i≠y} Z(x')_i, -confidence)
        real = logits.gather(1, y.view(-1, 1)).squeeze()
        other = (logits - 1e4 * F.one_hot(y, logits.size(1))).max(1)[0]
        f_loss = torch.clamp(real - other + confidence, min=0)
        
        dist = torch.norm(x_adv - x, p=2)
        loss = dist + c * f_loss
        
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        
        if f_loss.item() == 0 and dist.item() < best_dist:
            best_adv = x_adv.detach().clone()
            best_dist = dist.item()
    
    return best_adv
```

## Testing Strategy

```typescript
describe('AdversarialAttackSimulator', () => {
  describe('FGSM', () => {
    it('should generate adversarial examples within epsilon budget', async () => {
      const result = await simulator.fgsm(model, input, label, 0.031);
      expect(result.perturbationNorm).toBeLessThanOrEqual(0.031);
    });
    
    it('should achieve >50% success rate on undefended model', async () => {
      const report = await simulator.evaluateRobustness(model, testSet, labels, {
        method: 'fgsm', norm: 'Linf', epsilon: 0.031
      });
      expect(report.successfulAttacks / report.totalSamples).toBeGreaterThan(0.5);
    });
  });
  
  describe('PGD', () => {
    it('should be at least as strong as FGSM', async () => {
      const fgsmReport = await simulator.evaluateRobustness(model, testSet, labels, { method: 'fgsm', norm: 'Linf', epsilon: 0.031 });
      const pgdReport = await simulator.evaluateRobustness(model, testSet, labels, { method: 'pgd', norm: 'Linf', epsilon: 0.031, numSteps: 40 });
      expect(pgdReport.robustAccuracy).toBeLessThanOrEqual(fgsmReport.robustAccuracy);
    });
  });
  
  describe('AutoAttack', () => {
    it('should combine multiple attack strategies', async () => {
      const report = await simulator.evaluateRobustness(model, testSet, labels, { method: 'autoattack', norm: 'Linf', epsilon: 0.031 });
      expect(report).toBeDefined();
      expect(report.robustAccuracy).toBeGreaterThanOrEqual(0);
    });
  });
  
  describe('Transferability', () => {
    it('should measure cross-model attack transfer', async () => {
      const result = await simulator.transferAttack(sourceModel, targetModel, input, config);
      expect(result.success).toBeDefined();
    });
  });
});
```

## Acceptance Criteria

- [ ] FGSM, PGD, C&W, DeepFool, and AutoAttack implementations pass unit tests
- [ ] PGD achieves ≥90% attack success rate on undefended CIFAR-10 ResNet at ε=8/255
- [ ] C&W finds minimal perturbations within 2x of optimal
- [ ] AutoAttack ensemble correctly identifies non-robust samples
- [ ] Transferability analysis produces valid cross-model success rates
- [ ] Batch evaluation scales linearly with dataset size
- [ ] Attack results stored with full provenance (perturbation, logits, timing)
- [ ] All attacks respect perturbation budget constraints (Linf/L2/L1/L0)
- [ ] Documentation covers threat model assumptions and limitations

## References

- Goodfellow et al. (2015) — Explaining and Harnessing Adversarial Examples
- Madry et al. (2018) — Towards Deep Learning Models Resistant to Adversarial Attacks
- Carlini & Wagner (2017) — Towards Evaluating the Robustness of Neural Networks
- Croce & Hein (2020) — Reliable Evaluation of Adversarial Robustness with an Ensemble of Attacks
- Biggio & Roli (2018) — Wild Patterns: Ten Years After the Rise of Adversarial ML

---
*Sub-phase of [Phase 41 — Adversarial Robustness & Security Intelligence](../discussions/824)*
